src/python/pants/init/plugin_resolver.py

```diff
@@ -1,55 +1,98 @@
 import logging
-import os
-import shutil
 import site
-import uuid
-from typing import Iterable, Iterator, List, Optional, Type, TypeVar, cast
+from typing import Optional, TypeVar
 
-from pex import resolver
 from pex.interpreter import PythonInterpreter
-from pex.network_configuration import NetworkConfiguration
-from pkg_resources import Distribution, WorkingSet
+from pkg_resources import WorkingSet
 from pkg_resources import working_set as global_working_set
 
+from pants.backend.python.util_rules.pex import (
+    PexInterpreterConstraints,
+    PexRequest,
+    PexRequirements,
+    VenvPex,
+    VenvPexProcess,
+)
+from pants.core.util_rules.pants_environment import PantsEnvironment
+from pants.engine.collection import DeduplicatedCollection
+from pants.engine.internals.selectors import Params
+from pants.engine.internals.session import SessionValues
+from pants.engine.process import ProcessCacheScope, ProcessResult
+from pants.engine.rules import Get, QueryRule, collect_rules, rule
+from pants.init.engine_initializer import GraphScheduler
 from pants.option.global_options import GlobalOptions
-from pants.option.optionable import Optionable
 from pants.option.options_bootstrapper import OptionsBootstrapper
-from pants.option.scope import ScopeInfo
 from pants.option.subsystem import Subsystem
-from pants.python.python_repos import PythonRepos
-from pants.util.contextutil import temporary_dir
-from pants.util.dirutil import safe_concurrent_rename, safe_delete, safe_open
-from pants.util.memo import memoized_property
-from pants.version import PANTS_SEMVER
+from pants.util.logging import LogLevel
 
 logger = logging.getLogger(__name__)
 
 
 S = TypeVar("S", bound=Subsystem)
 
 
-class PluginResolver:
-    @staticmethod
-    def _is_wheel(path) -> bool:
-        return os.path.isfile(path) and path.endswith(".whl")
+class ResolvedPluginDistributions(DeduplicatedCollection[str]):
+    sort_input = True
+
+
+@rule
+async def resolve_plugins(global_options: GlobalOptions) -> ResolvedPluginDistributions:
+    """This rule resolves plugins using a VenvPex, and exposes the absolute paths of their dists.
+
+    NB: This relies on the fact that PEX constructs venvs in a stable location (within the
+    `named_caches` directory), but consequently needs to disable the process cache: see the
+    ProcessCacheScope reference in the body.
+    """
+    # TODO: Need a more bulletproof way to get constraints for the current interpreter and/or
+    # a way to specify a particular interpreter instead.
+    interpreter_constraints = PexInterpreterConstraints(
+        [f"=={PythonInterpreter.get().identity.version_str}"]
+    )
+    requirements = PexRequirements(sorted(global_options.options.plugins))
+    plugins_pex = await Get(
+        VenvPex,
+        PexRequest(
+            output_filename="pants_plugins.pex",
+            internal_only=True,
+            requirements=requirements,
+            interpreter_constraints=interpreter_constraints,
```

Comment:
If you read your gist closely you'll see:
```
... --constraint 3rdparty/python/constraints.txt toolchain.pants.plugin==0.5.0 ...
```
And that's bad because:
https://github.com/pantsbuild/pants/blob/9d02f990f080bedb99d4fc10fdb9dea72e02fc1b/3rdparty/python/constraints.txt#L32

Unfortantely, Pip provides a bad message here. Pex is tracking somehow improving this: https://github.com/pantsbuild/pex/issues/1200

The issue is here:
https://github.com/pantsbuild/pants/blob/e217ec9639b84c3ba0ffd458eeaf19df7f759b9a/src/python/pants/backend/python/util_rules/pex.py#L564-L565

You'll need a way to plumb "please don't use constraints".

Reply:
To be clear, the bigger problem is that Pants does not ship an API seperate from the binary. It needs to do so to solve this in general so the binary can pin everything but the api can float a bit. I suspect though, that when an api is extracted, it won't even have a dependency on setuptools. Nor much of anything else. And the problem goes poof: https://github.com/pantsbuild/pants/issues/9736

Reply:
Great, thanks!

Reply:
And really, my advice is wrong. The constraints really shouldn't be ignored. The plugin really should instead play the current dance and adjust its setuptools requirement to be compatible with Pants' requirement. I don't think its the case here, but it is the case in general that ignoring Pants' own constraints when installing plugins could lead to real incompatibilities at runtime.

Reply:
> The plugin really should instead play the current dance and adjust its setuptools requirement to be compatible with Pants' requirement.

Instead, Asher and I have been thinking that we should use transitive excludes so that Pants plugins do not depend on `pantsbuild.pants`. It creates a weird situation where using the plugin in `pantsbuild/pants` means we end up installing `pantsbuild.pants` into our venv, but also running from sources.

But this is blocked by a bug: https://github.com/pantsbuild/pants/issues/11469.

Reply:
> The constraints really shouldn't be ignored.

In this case they should, I think. The resolve for the code that is running "in the Pants process" is completely independent from the resolve for end user code, which is where `constraints.txt` comes into play. They don't mix.

Reply:
Ah yes, that's right. Pants dogfooding itself is confusing at times.